src/Microsoft.OData.Client/DataServiceContext.cs

```diff
@@ -3935,13 +4018,89 @@ private void DeleteObjectInternal(object entity, bool failIfInAddedState)
+                    List<string> dependsOnIdsAsChangeOrders;
+                    HashSet<string> dependsOnChangeSetIds;
+
+                    GetDependsOnChangeOrdersAndChangeSetIds(
+                        dependsOnIds,
+                        this.EntityTracker.Entities,
+                        this.Model,
+                        out dependsOnIdsAsChangeOrders,
+                        out dependsOnChangeSetIds);
+
+                    resource.DependsOnIds = dependsOnIdsAsChangeOrders;
+                    resource.DependsOnChangeSetIds = dependsOnChangeSetIds.ToList();
+                }
+
                 // Leave related links alone which means we can have a link in the Added
                 // or Modified state referencing a source/target entity in the Deleted state.
                 resource.State = EntityStates.Deleted;
+                resource.ChangeSetId = Guid.NewGuid().ToString();
                 this.entityTracker.IncrementChange(resource);
             }
         }
 
+        /// <summary>
+        /// Mark an existing object being tracked by the context for deletion.
+        /// </summary>
+        /// <param name="entity">entity to be mark deleted</param>
+        /// <param name="failIfInAddedState">If true, then an exception will be thrown if the entity is in the added state.</param>
+        /// <exception cref="ArgumentNullException">if entity is null</exception>
+        /// <exception cref="InvalidOperationException">if entity is not being tracked by the context, or if the entity is in the added state and <paramref name="failIfInAddedState"/> is true.</exception>
+        /// <remarks>
+        /// Existing objects in the Added state become detached if <paramref name="failIfInAddedState"/> is false.
+        /// </remarks>
+        private void DeleteObjectInternal(object entity, bool failIfInAddedState)
+        {
+            this.DeleteObjectInternal(entity, failIfInAddedState, null);
+        }
+
+        private static void GetDependsOnChangeOrdersAndChangeSetIds(
+            List<string> dependsOnIds,
+            IEnumerable<EntityDescriptor> entities,
+            ClientEdmModel model,
+            out List<string>  dependsOnIdsAsChangeOrders,
+            out HashSet<string> dependsOnChangeSetIds)
+        {
+            dependsOnIdsAsChangeOrders = new List<string>();
+            dependsOnChangeSetIds = new HashSet<string>();
+
+            foreach (string dependOnId in dependsOnIds)
+            {
+                foreach (EntityDescriptor entityDescriptor in entities)
+                {
+                    object entity = (object)entityDescriptor.Entity;
+                    string entityName = entity.GetType().FullName;
+                    string key = GetKeyPropertyName(model, entityName);
+                    object id = entity.GetType().GetProperty(key).GetValue(entity, null);
+                    if (dependOnId == id.ToString())

```

Comment:
@KenitoInc Should we do a case-insensitive string comparison here? To mitigate scenarios like:
- `dependsOnId = 0d599e52-ab31-4084-8b16-f5bebfcc70ef`, and
- `id = 0D599E52-AB31-4084-8B16-F5BEBFCC70EF`...

Reply:
We are no longer using this code section after design change